sdk/core/azure-core-http-okhttp/src/main/java/com/azure/core/http/okhttp/implementation/OkHttpAsyncResponse.java

```diff
@@ -111,6 +116,71 @@ public void writeBodyTo(OutputStream outputStream) throws IOException {
         StreamUtils.INSTANCE.transfer(responseBody.byteStream(), outputStream);
     }
 
+    @Override
+    public Mono<Void> writeBodyTo(AsynchronousFileChannel asynchronousFileChannel, long position) {
+        return Mono.defer(
+            () -> {
+                AtomicLong currentPosition = new AtomicLong(position);
+                byte[] buffer = new byte[8 * 1024];
+                ByteBuffer byteBuffer = ByteBuffer.wrap(buffer);
+                BufferedSource source = responseBody.source();
+                try {
+                    int read = source.read(buffer);
+                    byteBuffer.position(0);
+                    byteBuffer.limit(read);
+
+                    // TODO (kasobol-msft) consider using just reactor here.
+                    CompletableFuture<Void> completableFuture = new CompletableFuture<>();
+
+                    writeToChannel(
+                        asynchronousFileChannel, currentPosition, buffer, byteBuffer, source, completableFuture);
+
+                    return Mono.fromFuture(completableFuture);
+                } catch (IOException e) {
+                    return Mono.error(e);
+                }
+            }
+        ).doFinally(ignored -> close());
+    }
+
+    private void writeToChannel(
+        AsynchronousFileChannel asynchronousFileChannel, AtomicLong currentPosition,
+        byte[] buffer, ByteBuffer byteBuffer, BufferedSource source, CompletableFuture<Void> completableFuture) {
+        asynchronousFileChannel.write(byteBuffer, currentPosition.get(), byteBuffer,
+            new CompletionHandler<Integer, ByteBuffer>() {
+                @Override
+                public void completed(Integer result, ByteBuffer attachment) {
+                    currentPosition.addAndGet(result);
+
+                    if (byteBuffer.hasRemaining()) {
+                        // If the entire ByteBuffer hasn't been written send it to be written again until it completes.
+                        writeToChannel(
+                            asynchronousFileChannel, currentPosition, buffer, byteBuffer, source, completableFuture);
+                    } else {
+                        try {
+                            if (!source.exhausted()) {
+                                int read = source.read(buffer);
+                                byteBuffer.position(0);
+                                byteBuffer.limit(read);
+                                writeToChannel(
+                                    asynchronousFileChannel, currentPosition, buffer,
+                                    byteBuffer, source, completableFuture);
+                            } else {
+                                completableFuture.complete(null);
```

Comment:
Thinking about this design more, I think `FileChannel` makes more sense as this is effectively performing like a synchronous call as it "blocks" the reactive stream until the entire write operation completes. Given that I'd push to using `FileChannel`, unless there is a more nuanced reason why `AsynchronousFileChannel` should be used, as it'll simplify a lot of logic and potentially reduce the overall amount of thread hopping that occurs.

This statement should also echo true for the Reactor Netty implementation, and it'll give the added benefit of us not needing to handle retaining and releasing of Netty ByteBufs (which is another potential source of bugs) as the writing will block until it's complete and not introduce the `onNext` race condition called out earlier.

Reply:
When I was implementing this I missed the fact that `FileChannel` can write to arbitrary position in the file and I was under impression that we don't hava a choice. But I was wrong.
On the other hand this work isn't lost - this (or reactive) implementation is going to be useful for async apis that write to file.

Reply:
As for netty implementation - we will have to manage retain and release (and we have to introduce this for output stream writes). We shouldn't make blocking calls on nio threads (be it OutputStream.write or FileChannel.write) - they must land on Schedulers.elastic which means retain/release.
Imo we should pursue fully non-blocking (with async file channel) implementation for async apis.